### PR TITLE
[ISSUE #3197] NacosRestTemplate enhance

### PR DESCRIPTION
**Pull request by Maijh97**
Please do not create a Pull Request without creating an issue first.

## What is the purpose of the change

NacosRestTemplate enhance to meet the actual http client use
#3197 

## Brief changelog

- Add some parameter configuration of http client to the request method

- Add an interceptor for nacos resttemplate, can add some logic through the interceptor

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily:

* [ ] Make sure there is a Github issue filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
* [ ] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
* [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
* [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/alibaba/nacos/tree/master/test).
* [ ] Run `mvn -B clean package apache-rat:check findbugs:findbugs -Dmaven.test.skip=true` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.

### CODE REVIEW

**Comment by Maijh97** [comment]
please help to review this pr, thanks @KomachiSion

**Comment by KomachiSion** [comment]
We plan to release 1.3.1. We will stop merge PR before 1.3.1 released except code style PR and block issue PR.

I will review PR, if approved, I will merge it after 1.3.1 release.

File: common/src/main/java/com/alibaba/nacos/common/model/RequestHttpEntity.java [hunk 1]
```diff
@@ -31,19 +32,36 @@
...
+    private final HttpClientConfig httpClientConfig;
+    
     private final Query query;
     
     private Object body;
     
     public RequestHttpEntity(Header header, Query query) {
         handleHeader(header);
         this.query = query;
+        this.httpClientConfig = null;
```
Comment (KomachiSion) [inline common/src/main/java/com/alibaba/nacos/common/model/RequestHttpEntity.java:44]: change this constructer as `this(null, header, query, null)`?

File: common/src/main/java/com/alibaba/nacos/common/model/RequestHttpEntity.java [hunk 2]
```diff
@@ -31,19 +32,36 @@
...
+    }
+    
+    public RequestHttpEntity(HttpClientConfig httpClientConfig, Header header, Query query) {
+        handleHeader(header);
+        this.httpClientConfig = httpClientConfig;
+        this.query = query;
```
Comment (KomachiSion) [inline common/src/main/java/com/alibaba/nacos/common/model/RequestHttpEntity.java:50]: change this constructer as this(httpClientConfig, header, query, null)?

File: common/src/main/java/com/alibaba/nacos/common/model/RequestHttpEntity.java [hunk 3]
```diff
@@ -31,19 +32,36 @@
...
     }
     
     public RequestHttpEntity(Header header, Query query, Object body) {
         handleHeader(header);
         this.query = query;
         this.body = body;
+        this.httpClientConfig = null;
```
Comment (KomachiSion) [inline common/src/main/java/com/alibaba/nacos/common/model/RequestHttpEntity.java:57]: change this constructer as this(null, header, query, body)?

File: common/src/main/java/com/alibaba/nacos/common/http/client/InterceptingHttpClientRequest.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        this.httpClientRequest = httpClientRequest;
+        this.iterator = iterator;
+    }
+    
+    @Override
+    public HttpClientResponse execute(URI uri, String httpMethod, RequestHttpEntity requestHttpEntity)
+            throws Exception {
+        if (iterator.hasNext()) {
+            HttpClientRequestInterceptor nextInterceptor = iterator.next();
+            return nextInterceptor.intercept(uri, httpMethod, requestHttpEntity, this);
```
Comment (KomachiSion) [inline common/src/main/java/com/alibaba/nacos/common/http/client/InterceptingHttpClientRequest.java:47]: the Iterate method is too obscure, I suggest change this method.

**Comment by Maijh97** [comment]
I have adjusted the code.please help to review this pr again, thanks @KomachiSion

File: common/src/main/java/com/alibaba/nacos/common/http/client/NacosRestTemplate.java [hunk 5]
```diff
@@ -336,6 +460,16 @@ public NacosRestTemplate(HttpClientRequest requestClient) {
         }
     }
     
+    private HttpClientRequest requestClient() {
+        if (CollectionUtils.isNotEmpty(interceptors)) {
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug("Execute via interceptor");
```
Comment (KomachiSion) [inline common/src/main/java/com/alibaba/nacos/common/http/client/NacosRestTemplate.java:466]: Should we print the interceptor name or class name?

Comment (Maijh97) [reply common/src/main/java/com/alibaba/nacos/common/http/client/NacosRestTemplate.java]: okay

**Pull request merged**